client.go

```diff
@@ -211,6 +357,140 @@ func (c *Client) monitorChannel(ctx context.Context) {
+	// if err := c.sechan.Close(); err != nil && err != io.EOF {
+	// 	return nil, nil, err
+	// }
+	conn, err := uacp.Dial(ctx, c.endpointURL)
+	if err != nil {
+		return nil, nil, err
+	}
+	sechan, err := uasc.NewSecureChannel(c.endpointURL, conn, c.cfg)
+	if err != nil {
+		_ = conn.Close()
+		return nil, conn, err
+	}
+	return sechan, conn, nil
+}
+
+func (c *Client) repairSessions() error {
+	for s := range c.sessions {
+		if err := c.repairSession(s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Client) repairSession(s *Session) error {
+
+	if _, err := c.DetachSession(); err != nil {
+		return err
+	}
+
+	debug.Printf("Trying to reactivate existing session")
+
+	err := c.ActivateSession(s)
+	if err != nil {
+		return errors.Errorf("Session reactivation failed")
+	} else {
+		debug.Printf("Existing session reactivated trying to repair subscriptions")
+		if err := c.repairSubscriptions(); err != nil {
+			return err
+		}
+		debug.Printf("Subscriptions repaired")
+	}
+
+	return nil
+}
+
+func (c *Client) SubscriptionIDs() []uint32 {
+	subscriptionIDs := []uint32{}
+	c.subMux.Lock()
+	for key := range c.subscriptions {
+		subscriptionIDs = append(subscriptionIDs, key)
+	}
+	c.subMux.Unlock()
+	return subscriptionIDs
+}
+
+func (c *Client) repairSubscriptions(subscriptionIDs ...uint32) error {
+
+	if subscriptionIDs == nil {
+		subscriptionIDs = c.SubscriptionIDs()
+	}
+
+	c.subMux.RLock()
+	defer c.subMux.RUnlock()
+
+	for _, subID := range subscriptionIDs {
+		sub, ok := c.subscriptions[subID]
+		if !ok {
+			return errors.Errorf("Invalid SubscriptionID")
+		}
+		if err := c.repairSubscription(sub); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) repairSubscription(sub *Subscription) error {
+	debug.Printf("RepairSubscription  for SubscriptionId %d", sub.SubscriptionID)
+	if status, err := c.subscriptionRepublish(sub); err != nil {
+		switch status {
+		case ua.StatusBadSessionIDInvalid:
+			return err
+		case ua.StatusBadSubscriptionIDInvalid:
+			debug.Printf("Republish failed, subscriptionId is not valid anymore on server side.")
+			// return sub.recreateSubscriptionAndMonitoredItem()
+			return errors.Errorf("Republish failed, subscriptionId is not valid anymore on server side.")
+		}
+	}
+	return nil
+}
+
+func (c *Client) subscriptionRepublish(sub *Subscription) (ua.StatusCode, error) {
+	isDone := false
+
+	for !isDone {
+		request := &ua.RepublishRequest{
+			SubscriptionID:           sub.SubscriptionID,
+			RetransmitSequenceNumber: sub.lastSequenceNumber + 1,
+		}
+
+		debug.Printf("Republish Request for subscription %d retransmitSequenceNumber=%d",
+			request.SubscriptionID,
+			request.RetransmitSequenceNumber,
+		)
+
+		if c.sessionIsClosed() {
+			debug.Printf("Publish engine republish aborted")
+			isDone = true
+			continue
+		}
+		var res *ua.RepublishResponse
+		var err error
+
+		res, err = sub.republish(request)
```

Comment:
```go
res, err := sub.republish(req)
```

This looks fishy since you check `res` before you check `err`. Please check this and try to refactor with `switch`. Nested code is harder to read.